<a href="https://colab.research.google.com/github/CRPeace/Dojo-food_items_assignment/blob/main/C_Peace_Food_Items_Predictions_rev_12_17_22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Food Item Predictions
##### Cameron Peace

##### Assignment Checklist


1. [x] How many rows and columns?
2. [x] What are the datatypes of each variable?
3. [x] Are there duplicates? If so, drop any duplicates.
4. [x] Identify missing values.
5. [x] Decide on how to address the missing values and do it! (This requires your judgement, so explain your choice).
6. [x] Confirm that there are no missing values after addressing them.
7. [x] Find and fix any inconsistent categories of data (example: fix cat, Cat, and cats so that they are consistent).
8. [x] For any numerical columns, obtain the summary statistics of each (min, max, mean).

##### Data Dictionary


* Variable Name: --	Description
* Item_Identifier: --	Unique product ID
* Item_Weight: --	Weight of product
* Item_Fat_Content: --	Whether the product is low fat or regular
* Item_Visibility: --	The percentage of total display area of all products in a store allocated to the particular product
* Item_Type: --	The category to which the product belongs
* Item_MRP: --	Maximum Retail Price (list price) of the product
* Outlet_Identifier: --	Unique store ID
* Outlet_Establishment_Year: --	The year in which store was established
* Outlet_Size: --	The size of the store in terms of ground area covered
* Outlet_Location_Type: --	The type of area in which the store is located
* Outlet_Type: --	Whether the outlet is a grocery store or some sort of supermarket
* Item_Outlet_Sales: --	Sales of the product in the particular store. This is the target variable to be predicted.

### <font color='green'> Loading Data </font>

In [1]:
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [2]:
df = pd.read_csv('/content/sales_predictions.csv')

In [3]:
display(df.head(), round(df.describe(), 2), df.info(), df.isna().sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
count      7060.00          8523.00   8523.00                    8523.00   
mean         12.86             0.07    140.99                    1997.83   
std           4.64             0.05     62.28                       8.37   
min           4.56             0.00     31.29                    1985.00   
25%           8.77             0.03     93.83                    1987.00   
50%          12.60             0.05    143.01                    1999.00   
75%          16.85             0.09    185.64                    2004.00   
max          21.35             0.33    266.89                    2009.00   

       Item_Outlet_Sales  
count            8523.00  
mean             2181.29  
std              1706.50  
min                33.29  
25%               834.25  
50%              1794.33  
75%              3101.30  
max             13086.96

None

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

### <font color='green'>Data Cleaning</font>

##### 1.
1. How many rows and columns? <font color='blue'> There are 8523 rows and 12 columns in this dataset </font>
2. What are the datatypes of each variable? <font color='blue'> See below, objects, floats and 1 int dtype.  All look consistent with value types </font>
3. Are there duplicates? If so, drop any duplicates. <font color='blue'> There are no duplicate rows in this dataset </font>

In [4]:
df.shape

(8523, 12)

In [5]:
df.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

In [6]:
print(f'There are {df.duplicated().sum()} duplicate rows in the dataset')
df.loc[df.duplicated(), :]

There are 0 duplicate rows in the dataset


Empty DataFrame
Columns: [Item_Identifier, Item_Weight, Item_Fat_Content, Item_Visibility, Item_Type, Item_MRP, Outlet_Identifier, Outlet_Establishment_Year, Outlet_Size, Outlet_Location_Type, Outlet_Type, Item_Outlet_Sales]
Index: []

##### 2.
4. Identify missing values. <font color='blue'> There are 1463 missing values in "Item_Weight" and 2410 missing values in "Outlet_Size". </font>
5. Decide on how to address the missing values and do it! (This requires your judgement, so explain your choice).  
<font color='blue'>   * For the missing values in "Item_Weight":  They are all confined to 2 specific stores.  It would depend on what kinds of questions we're trying to answer, but for the purposes of this exercise, I'm assuming that we're interested in summary statistics for weight and so I would drop nulls as opposed to replacing them with zero or imputing (so we get accurate numbers for our summary statistics).  We could also just fill them all with 0 and add a dummy variable colum "missing_weight" to keep them in the dataset while making it easy to filter them out as needed. </font>
<font color='green'> * For the missing values in "Outlet_Size" it makes most sense to me to fill it with a "missing value" string, since the dtype is already an object. </font>
6. Confirm that there are no missing values after addressing them. <font color='blue'> Confirmed in last line of section </font>

In [7]:
# Looks like we have missing values in the "Item_Weight" and "Outlet_Size" columns
df.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

In [8]:
# taking a look at other values in the rows with missing weight in case there is a pattern.
df[df['Item_Weight'].isna()]
df[df['Item_Weight'].isna()]['Outlet_Identifier'].value_counts()

OUT027    935
OUT019    528
Name: Outlet_Identifier, dtype: int64

In [9]:
# looks like the missing weights are confined to only 2 stores: 'OUT027' and 'OUT019'
# would consult with subject matter expert, but seems simplest to drop those stores from the dataset since we're missing data from them
# dropping 2 out of 10 stores does take a bite out of the dataset though. We'll lose around 17 percent of the data here.
df['Outlet_Identifier'].value_counts(normalize=True)  

OUT027    0.109703
OUT013    0.109351
OUT049    0.109117
OUT046    0.109117
OUT035    0.109117
OUT045    0.108999
OUT018    0.108882
OUT017    0.108647
OUT010    0.065118
OUT019    0.061950
Name: Outlet_Identifier, dtype: float64

In [10]:
# This filters out any stores that are not OUT027 or OUT019, since in this case all nulls are coming from these two stores
# Although it makes just as much sense to drop nulls via the "Item_Weight" column \
#     I prefer this way since it makes it clear that data from two stores are problematic and removes all values.
cleaned = df.copy()
cleaned = df[~df['Outlet_Identifier'].isin(['OUT027', 'OUT019'])]

In [11]:
# verifying that we're cleaned for "Item_Weight"
print(cleaned.shape)
print(cleaned.isna().sum())

(7060, 12)
Item_Identifier                 0
Item_Weight                     0
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64


In [12]:
# Now filling all nulls from "Outlet_Size" with 'missing value'
cleaned.loc[:, 'Outlet_Size'] = cleaned.loc[:, 'Outlet_Size'].fillna('missing value')

/usr/local/lib/python3.8/dist-packages/pandas/core/indexing.py:1773: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


In [13]:
# Verifying that there are no more missing values in the whole dataset
cleaned.isna().sum()

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

##### 3.
7. Find and fix any inconsistent categories of data (example: fix cat, Cat, and cats so that they are consistent). <font color='blue'> There were several iterations of low fat and regular in the "Item_Fat_Content" column.  I fixed those to either "Low Fat" or "Regular".  For the "Outlet_Size" column, the value "High" doesn't seem to belong amid "Small" and "Medium" ("Large" feels more appropriate), but would ask subject matter expert. I didn't see any other problems.  Some columns should be turned into categorical ordinal values, but that is probably not necessary right now.  </font>
8. For any numerical columns, obtain the summary statistics of each (min, max, mean). <font color='blue'> I used .describe() to view summary statistics, including all in case something jumped out from one of the other columns.  Looking good as far as I can tell. </font>

In [14]:
cleaned.columns # printing out column names to copy
# finding those I want to take a closer look at
my_columns = ['Item_Fat_Content',
       'Item_Type', 'Outlet_Identifier',
       'Outlet_Establishment_Year', 'Outlet_Size', 'Outlet_Location_Type',
       'Outlet_Type']
for i in my_columns:
  print(i.upper(), '****\n', cleaned[i].value_counts(dropna=False))
  print('**************\n\n')

ITEM_FAT_CONTENT ****
 Low Fat    4222
Regular    2388
LF          260
reg         106
low fat      84
Name: Item_Fat_Content, dtype: int64
**************


ITEM_TYPE ****
 Fruits and Vegetables    1019
Snack Foods               988
Household                 759
Frozen Foods              718
Dairy                     566
Canned                    539
Baking Goods              536
Health and Hygiene        430
Soft Drinks               374
Meat                      337
Breads                    204
Hard Drinks               183
Others                    137
Starchy Foods             130
Breakfast                  89
Seafood                    51
Name: Item_Type, dtype: int64
**************


OUTLET_IDENTIFIER ****
 OUT013    932
OUT049    930
OUT046    930
OUT035    930
OUT045    929
OUT018    928
OUT017    926
OUT010    555
Name: Outlet_Identifier, dtype: int64
**************


OUTLET_ESTABLISHMENT_YEAR ****
 1987    932
1999    930
1997    930
2004    930
2002    929
2009    928
2007 

In [15]:
#Fixing the "Item_Fat_Content" column, trying it with a custom function just to give it a whirl.
def fix_fat_label(label):
  if  label== 'low fat':
    return 'Low Fat'
  elif label == 'LF':
    return 'Low Fat'
  elif label == 'reg':
    return 'Regular'
  else:
    return label
    
cleaned.loc[:,'Item_Fat_Content'] = cleaned.loc[:,'Item_Fat_Content'].apply(fix_fat_label)

In [16]:
# checking things
cleaned['Item_Fat_Content'].value_counts()

Low Fat    4566
Regular    2494
Name: Item_Fat_Content, dtype: int64

In [17]:
cleaned.describe(include='all').T

count unique                    top  freq  \
Item_Identifier              7060   1555                  FDF22     8   
Item_Weight                7060.0    NaN                    NaN   NaN   
Item_Fat_Content             7060      2                Low Fat  4566   
Item_Visibility            7060.0    NaN                    NaN   NaN   
Item_Type                    7060     16  Fruits and Vegetables  1019   
Item_MRP                   7060.0    NaN                    NaN   NaN   
Outlet_Identifier            7060      8                 OUT013   932   
Outlet_Establishment_Year  7060.0    NaN                    NaN   NaN   
Outlet_Size                  7060      4          missing value  2410   
Outlet_Location_Type         7060      3                 Tier 2  2785   
Outlet_Type                  7060      3      Supermarket Type1  5577   
Item_Outlet_Sales          7060.0    NaN                    NaN   NaN   

                                  mean          std     min       25%  \
Item_Identifier                    NaN          NaN     NaN       NaN   
Item_Weight                  12.857645     4.643456   4.555   8.77375   
Item_Fat_Content                   NaN          NaN     NaN       NaN   
Item_Visibility               0.063963     0.048625     0.0  0.026768   
Item_Type                          NaN          NaN     NaN       NaN   
Item_MRP                    141.240683    62.411888   31.49   94.1436   
Outlet_Identifier                  NaN          NaN     NaN       NaN   
Outlet_Establishment_Year  2000.490935     6.588984  1987.0    1997.0   
Outlet_Size                        NaN          NaN     NaN       NaN   
Outlet_Location_Type               NaN          NaN     NaN       NaN   
Outlet_Type                        NaN          NaN     NaN       NaN   
Item_Outlet_Sales          2118.626808  1533.445842   33.29  922.7988   

                                 50%       75%        max  
Item_Identifier                  NaN       NaN        NaN  
Item_Weight                     12.6     16.85      21.35  
Item_Fat_Content                 NaN       NaN        NaN  
Item_Visibility             0.052493  0.092774    0.31109  
Item_Type                        NaN       NaN        NaN  
Item_MRP                    142.7299   186.024   266.8884  
Outlet_Identifier                NaN       NaN        NaN  
Outlet_Establishment_Year     2002.0    2007.0     2009.0  
Outlet_Size                      NaN       NaN        NaN  
Outlet_Location_Type             NaN       NaN        NaN  
Outlet_Type                      NaN       NaN        NaN  
Item_Outlet_Sales          1789.6704  2966.139  10256.649

### <font color='green'>Exploratory Visuals</font>

### <font color='green'>Explanatory Visuals</font>